slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,301 @@
+
+	"k8s.io/apiserver/pkg/storage"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+
+	kubeletevents "k8s.io/kubernetes/pkg/kubelet/events"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// PodE2EStartupLatencyKey is a key for pod startup latency monitoring metric.
+	PodE2EStartupLatencyKey = "slomonitor_pod_e2e_startup_latency_seconds"
+)
+
+var (
+	// PodE2EStartupLatency is a prometheus metric for monitoring pod startup latency.
+	PodE2EStartupLatency = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    PodE2EStartupLatencyKey,
+			Help:    "Pod e2e startup latencies in milliseconds, without image pull times",
+			Buckets: prometheus.LinearBuckets(0.5, 0.25, 50),
+		},
+	)
+)
+
+var registerMetrics sync.Once
+
+// Register registers all prometheus metrics.
+func Register() {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(PodE2EStartupLatency)
+	})
+}
+
+// PodStartupMilestones keeps all milestone timestamps from Pod creation.
+type PodStartupMilestones struct {
+	created         time.Time
+	startedPulling  time.Time
+	finishedPulling time.Time
+	observedRunning time.Time
+}
+
+// PodStartupLatencyDataMonitor monitors pod startup latency and exposes prometheus metric.
+type PodStartupLatencyDataMonitor struct {
+	sync.Mutex
+	kubeClient     clientset.Interface
+	PodStartupData map[string]PodStartupMilestones
+	startPodRV     uint64
+	startEventRV   uint64
+}
+
+// NewPodStartupLatencyDataMonitor creates a new PodStartupLatencyDataMonitor from a given client.
+func NewPodStartupLatencyDataMonitor(c clientset.Interface) *PodStartupLatencyDataMonitor {
+	return &PodStartupLatencyDataMonitor{
+		kubeClient:     c,
+		PodStartupData: map[string]PodStartupMilestones{},
+	}
+}
+
+// Run starts a PodStartupLatencyDataMonitor: it creates all watches, populates PodStartupData and updates
+//  PodE2EStartupLatency metric
+func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) {
+	// When starting we need to get a 'current' ResourceVersion, so that
+	podsList, err := pm.kubeClient.Core().Pods(v1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		panic(err)
+	}
+	startPodRV, err := storage.ParseListResourceVersion(podsList.ResourceVersion)
+	if err != nil {
+		panic(err)
+	}
+	glog.Infof("Initial Pod RV: %v", startPodRV)
+	pm.startPodRV = startPodRV
+
+	eventList, err := pm.kubeClient.Core().Events(v1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		panic(err)
+	}
+	startEventRV, err := storage.ParseListResourceVersion(eventList.ResourceVersion)
+	if err != nil {
+		panic(err)
+	}
+	glog.Infof("Initial Event RV: %v", startEventRV)
+	pm.startEventRV = startEventRV
+
+	controller := NewWatcherWithHandler(&cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			obj, err := pm.kubeClient.Core().Pods(v1.NamespaceAll).List(options)
```

Comment:
Actually I need to, as it's possible that something was created just before monitor startup, and some of the events will come when it's already running and will be normally accounted for. Which we don't want to do, which means that I need to populate the data for them.